broadlink/climate.py

```diff
@@ -235,3 +238,217 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
             input_payload.append(int(weekend[i]["temp"] * 2))
 
         self.send_request(input_payload)
+
+
+class hvac(device):
+    """Controls a HVAC.
+
+    Supported models:
+    - Tornado SMART X SQ series.
+    """
+
+    @unique
+    class Mode(IntEnum):
+        """Enumerates modes."""
+        AUTO = 0
+        COOL = 1
+        DRY = 2
+        HEAT = 3
+        FAN = 4
+
+    @unique
+    class Speed(IntEnum):
+        """Enumerates fan speed."""
+        HIGH = 1
+        MID = 2
+        LOW = 3
+        AUTO = 5
+
+    @unique
+    class Preset(IntEnum):
+        """Enumerates presets."""
+        NORMAL = 0
+        TURBO = 1
+        MUTE = 2
+
+    @unique
+    class SwHoriz(IntEnum):
+        """Enumerates horizontal swing."""
+        ON = 0
+        OFF = 7
+
+    @unique
+    class SwVert(IntEnum):
+        """Enumerates vertical swing."""
+        ON = 0
+        POS1 = 1
+        POS2 = 2
+        POS3 = 3
+        POS4 = 4
+        POS5 = 5
+        OFF = 7
+
+    def __init__(self, *args, **kwargs):
+        device.__init__(self, *args, **kwargs)
+        self.type = "HVAC"
+
+    def _crc(self, data: bytes) -> int:
+        """Calculate CRC of a byte object."""
+        s = sum([v if i % 2 == 0 else v << 8 for i, v in enumerate(data)])
+        # trim the overflow and add it to smallest bit
+        s = (s & 0xffff) + (s >> 16)
+        return (0xffff - s)  # TODO: fix: we can't return negative values
```

Comment:
I might be misunderstanding what you mean.
![Screen Shot 2021-05-02 at 14 23 06](https://user-images.githubusercontent.com/3087025/116811461-23ba1e00-ab52-11eb-9496-ba15ce35c49f.png)
But if you mean be negative like in the screenshot, I think it'd be okay just to check that the packet is smaller than 2^18B.

Reply:
Yes, overflow is unlikely as we are not dealing with packets of this size. `& 0xFFFF` should be enough as a safeguard:

```python
     return (0xffff - s) & 0xffff
```

Reply:
I just found the polynomial, now we can do:
```python3
crc = CRC16.calculate(payload[0x02:p_len], polynomial=0x9BE4)
```